modules/virt-creating-a-service-from-a-running-virtual-machine.adoc

```diff
@@ -0,0 +1,145 @@
+machine is running, create a service object to expose the service. Connect
+to the service by specifying the service name.
+
+The service exposes the host name of the virtual machine to the internal cluster,
+using the `ClusterIP` service, or externally, using the `NodePort` or `LoadBalancer`
+service.
+
+This procedure presents an example of how to create and expose a service object
+of `type: ClusterIP` as a virtual machine-backed service.
+
+[NOTE]
+====
+ClusterIP is the default service `type`, if the service `type` is not specified.
+====
+
+.Procedure
+
+. Edit the `VirtualMachine` YAML configuration file to include the
+label `special: key:`.
++
+
+[NOTE]
+====
+Labels on a `VirtualMachine` are passed through to the pod. The labels on
+the `VirtualMachine`, for example `special: key:`, must match the labels on
+the Service object.
+====
+
++
+
+[source,yaml]
+----
+apiVersion: kubevirt.io/v1alpha3
+kind: VirtualMachine
+metadata:
+  name: vmi-ephemeral
+  labels:
+    special: key <1>
+spec:
+  domain:
+    devices:
+      disks:
+      - disk:
+          bus: virtio
+        name: containerdisk
+    resources:
+      requests:
+        memory: 64M
+  volumes:
+  - name: containerdisk
+    containerDisk:
+      image: kubevirt/fedora-cloud-container-disk-demo:latest
+----
+<1> The `special: key:` label
++
+
+. Edit the YAML to create and expose the `Service` object, connecting by
+specifying the service name.
+
+.. Add `type: ClusterIP` for a
+ClusterIP service. Add `targetPort: 22`, exposing the service on SSH port `22`.
++
+To create and expose other types of services, such as `NodePort` and `LoadBalancer`, replace
+`type: ClusterIP` with `type: NodePort` or `type: LoadBalancer`, as appropriate,
+in the Service object definition.
++
+.. Specify a `name` for the service you are creating and exposing.
++
+
+[source,yaml]
+----
+apiVersion: v1
+kind: Service
+metadata:
+  name: vmiservice <1>
+spec:
+  ports:
+  - port: 27017
+    protocol: TCP
+    targetPort: 22 <2>
+  selector:
+    special: key
+  type: ClusterIP <3>
+----
+<1> Service name
+<2> Exposes SSH `targetPort` 22
+<3> Defines Service `type: ClusterIP`
+
++
+. Create the `ClusterIP` service using `kubectl`. Specify the `name`
+you defined when you created the `Service` object as the `service_name`.
++
+
+[source,terminal]
+----
+$ kubectl create -f <service_name>.yaml
+----
+
++
+. Start the virtual machine. If the virtual machine is already running, restart it.
++
+
++
+. Query the Service object to verify it is available and is configured with type `ClusterIP`.
++
+
+[source, terminal]
+----
+$ kubectl get service
+----
++
+
+.Example output
+[source, terminal]
+----
+NAME        TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)     AGE
+vmiservice  ClusterIP   172.30.3.149   <none>        27017/TCP   2m
+----
++
+
+. Connect to the `VirtualMachine` using the service IP and service port.
++
+
+[source,terminal]
+----
+$ ssh <host_name>@<node_IP> -p 27017
```

Comment:
1. Username (the one which is used to login to the VM), not host_name.
2. Which node's IP? The user should be instructed to find which node hosts the VM (can be achieved with `oc get vmi`), and then find that node's IP (e.g. run `oc get node <node-name> -o wide`, and take the IP address from the `INTERNAL-IP` column).

Reply:
I expanded this and snagged some example output from the `get node` command:

. Connect to the `VirtualMachine` using the service IP and service port.
* Exposing a virtual machine's ssh port through a `ClusterIP` service enables
connection to a virtual machine within the cluster, either through
another pod or virtual machine, for example.
* `NodePort` and `LoadBalancer` services enable connection
from outside of a cluster, as long as you can tunnel
traffic to the cluster's node IP addresses.
.. Find the node that hosts the virtual machine. List all nodes with information
about a project’s pod deployment by running:
+

[source,terminal]
----
$ oc get node <node_name> -o wide
----
+

.Example output
[source,terminal]
----
NAME                           STATUS    ROLES     AGE       VERSION           INTERNAL-IP      OS-IMAGE                            KERNEL-VERSION          CONTAINER-RUNTIME
ip-172-18-0-199.ec2.internal   Ready     infra     1d        v1.10.0+b81c8f8   54.172.185.130   Fedora cloud container disk demo    3.10.0-862.el7.x86_64   docker://1.13.1
----
+

.. Connect using the service IP and service port. In the `ssh` command, replace
`<node_IP>` with the IP address in the INTERNAL-IP column from the previous `get node`
command output. In this example, the `INTERNAL-IP` is `54.172.185.130`.

+

[source,terminal]
----
$ ssh <user_name>@<node_IP> -p 27017
----

Tagging @yossisegev and @phoracek  for verfication.